### PR TITLE
Handle NaN in array_with_unit_to_datetime

### LINKED ISSUE #48610: BUG: pandas.to_datetime fails to handle numpy.nan on riscv64 due to dependency on undefined behaviour
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas
import numpy
print(pandas.to_datetime(numpy.nan, unit="s"))
```

### Issue Description

> Converting a value of floating type to integer type which is out of range for the integer type is undefined, see 6.3.1.4 Real floating and integer.

> You can use gcc92.fsffrance.org for your tests if don't have your own hardware, or use qemu with the images from   https://download.opensuse.org/ports/riscv/tumbleweed/images/.

```
$ python3 -c 'import pandas
import numpy
print(pandas.to_datetime(numpy.nan, unit="s"))'
Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "/usr/lib64/python3.10/site-packages/pandas/core/tools/datetimes.py", line 1078, in to_datetime
    result = convert_listlike(np.array([arg]), format)[0]
  File "/usr/lib64/python3.10/site-packages/pandas/core/tools/datetimes.py", line 357, in _convert_listlike_datetimes
    return _to_datetime_with_unit(arg, unit, name, tz, errors)
  File "/usr/lib64/python3.10/site-packages/pandas/core/tools/datetimes.py", line 530, in _to_datetime_with_unit
    arr, tz_parsed = tslib.array_with_unit_to_datetime(arg, unit, errors=errors)
  File "pandas/_libs/tslib.pyx", line 266, in pandas._libs.tslib.array_with_unit_to_datetime
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: cannot convert input with unit 's'

```

### Expected Behavior

No error.

### Installed Versions

<details>

/usr/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

INSTALLED VERSIONS
------------------
commit           : e8093ba372f9adfe79439d90fe74b0b5b6dea9d6
python           : 3.10.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 6.0.0-rc5-38-default
Version          : #1 SMP Mon Sep 12 15:18:20 UTC 2022 (005845a)
machine          : riscv64
processor        : riscv64
byteorder        : little
LC_ALL           : None
LANG             : de_DE.UTF-8
LOCALE           : de_DE.UTF-8

pandas           : 1.4.3
numpy            : 1.21.6
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 63.2.0
pip              : 22.0.4
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : 1.3.5
brotli           : 1.0.9
fastparquet      : None
fsspec           : None
gcsfs            : None
markupsafe       : None
matplotlib       : None
numba            : None
numexpr          : 2.8.3
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by andreas-schwab**
closes https://github.com/pandas-dev/pandas/issues/48610

Missing values in arrays can either be NaT (for integer types) or NaN (for
floating types).  Make sure to also handle the latter.  This fixes all
failing tests.

### CODE REVIEW

**Comment by mroeschke** [comment]
Does this fix an existing Github issue? Would need tests as well

**Comment by andreas-schwab** [comment]
The existing tests already cover it perfectly.

**Comment by mroeschke** [comment]
You mentioned 

> This fixes all failing tests

What tests were failing?

**Comment by andreas-schwab** [comment]
https://build.opensuse.org/package/live_build_log/openSUSE:Factory:RISCV/python-pandas:test-py310/standard/riscv64

=========================== short test summary info ============================
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_convert_dates_infer[trade_time]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_convert_dates_infer[date]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_convert_dates_infer[datetime]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_convert_dates_infer[sold_at]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_convert_dates_infer[modified]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_convert_dates_infer[timestamp]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_convert_dates_infer[timestamps]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_date_unit[s]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_date_unit[ms]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_date_unit[us]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_to_json_from_json_columns_dtypes[split]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_to_json_from_json_columns_dtypes[records]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_to_json_from_json_columns_dtypes[index]
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_to_json_from_json_columns_dtypes[columns]
FAILED pandas/tests/tools/test_to_datetime.py::TestToDatetimeUnit::test_to_datetime_unit_with_nulls[nan]
= 15 failed, 143030 passed, 9020 skipped, 1047 xfailed, 6 xpassed, 18 warnings in 16656.29s (4:37:36) =

**Comment by mroeschke** [comment]
So this is related to https://github.com/pandas-dev/pandas/issues/48610?

**Comment by andreas-schwab** [comment]
Yes.

**Review by mroeschke — COMMENTED** [review]
Looks okay. cc @jbrockmendel if any further comments.

File: pandas/_libs/tslib.pyx [hunk 1]
```diff
@@ -301,6 +301,9 @@ def array_with_unit_to_datetime(
             iresult = values.astype("i8", copy=False)
             # fill missing values by comparing to NPY_NAT
             mask = iresult == NPY_NAT
+            # also handle NaN
+            if values.dtype.kind == "f":
+                mask |= values != values
```
Comment (jbrockmendel) [inline pandas/_libs/tslib.pyx:306]: on the relevant hardware, what do you get for iresult when values is [np.nan]?

Comment (andreas-schwab) [reply pandas/_libs/tslib.pyx:306]: $ python3
Python 3.10.7 (main, Sep 11 2022, 08:41:56) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.array([np.nan]).astype("i8")
array([9223372036854775807])

Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:306]: @seberg is getting 9223372036854775807 instead of 9223372036854775808 weird or is it just me?

Comment (andreas-schwab) [reply pandas/_libs/tslib.pyx:306]: 42 would also be a valid result.

Comment (seberg) [reply pandas/_libs/tslib.pyx:306]: NumPy has no guarantees for `NaN -> int64` casts currently! (Although they are better implemented, in a sense).  So what you get is completely off-bets.  For example on arm, you may get 0!
And that breaks `mask = iresult == NPY_NAT`?  There should be tests failing here I would think?

However, NumPy *does* have guarantees for `NaN -> datetime/timedelta` casts, so going via `i8` seems dubious in this case?  The cast triggers undefined behavior in C.

I am not sure what you mean with:
> 42 would also be a valid result.

@andreas-schwab?

Comment (andreas-schwab) [reply pandas/_libs/tslib.pyx:306]: No sure why you ask, you already gave the answer above.

Comment (seberg) [reply pandas/_libs/tslib.pyx:306]: Misread the code.  In any case, NumPy handles the cast correctly for datetime64, even if it uses a slower/weirder code path (currently, this is very much fixable!).  So the question is wether it doesn't make more sense to cast to datetime64.

Further, the above may cause floating-point warnings in newer NumPy versions, which is something you probably don't want to see!

**Comment by andreas-schwab** [comment]
$ python3
Python 3.10.7 (main, Sep 11 2022, 08:41:56) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.array([np.nan]).astype("i8")
array([9223372036854775807])

File: pandas/_libs/tslib.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:301]: this is fine.  can you add a comment pointing back to this PR and a few words about why we're doing this

**Comment by MarcoGorelli** [comment]
Would https://github.com/pandas-dev/pandas/pull/50183/files also address the issue?

**Comment by andreas-schwab** [comment]
Yes, it does.

https://build.opensuse.org/project/monitor/home:Andreas_Schwab:riscv:python

**Review by MarcoGorelli — APPROVED** [review]
Looks fine to me, and I think the comment's clear enough now. I've just merged with upstream/main to get CI green

File: pandas/_libs/tslib.pyx [hunk 2]
```diff
@@ -300,6 +300,10 @@ def array_with_unit_to_datetime(
             iresult = values.astype("i8", copy=False)
             # fill missing values by comparing to NPY_NAT
             mask = iresult == NPY_NAT
+            # Trying to Convert NaN to integer results in undefined
+            # behaviour, so handle it explicitly
```
Comment (MarcoGorelli) [inline pandas/_libs/tslib.pyx:304]:
```suggestion
            # behaviour, so handle it explicitly (see GH #48705)
```

**Comment by MarcoGorelli** [comment]
thanks @andreas-schwab

**Pull request merged**